src/main/java/de/srendi/advancedperipherals/common/addons/appliedenergistics/AECraftJob.java

```diff
@@ -0,0 +1,309 @@
+            calculationNotSuccessful = true;
+            return;
+        }
+
+        futureJob = craftingService.beginCraftingCalculation(world, this.simulationRequester, toCraft, amount, CalculationStrategy.REPORT_MISSING_ITEMS);
+        fireEvent(false, StatusConstants.CALCULATION_STARTED);
+    }
+
+    public void maybeCraft() {
+        if (startedCrafting || futureJob == null || !futureJob.isDone()) {
+            return;
+        }
+        ICraftingPlan job;
+
+        try {
+            job = futureJob.get();
+        } catch (ExecutionException | InterruptedException ex) {
+            AdvancedPeripherals.debug("Tried to get job, but job calculation is not done. Should be done.", org.apache.logging.log4j.Level.ERROR);
+            ex.printStackTrace();
+            fireEvent(true, StatusConstants.UNKNOWN_ERROR);
+            return;
+        }
+
+        if (job == null) {
+            AdvancedPeripherals.debug("Job is null, should not be null.", org.apache.logging.log4j.Level.ERROR);
+            fireEvent(true, StatusConstants.UNKNOWN_ERROR);
+            return;
+        }
+        this.currentJob = job;
+
+        KeyCounter missing = job.missingItems();
+        if (!missing.isEmpty()) {
+            fireEvent(true, StatusConstants.MISSING_ITEMS);
+            calculationNotSuccessful = true;
+            return;
+        }
+
+        IGrid grid = node.getGrid();
+
+        ICraftingService craftingService = grid.getService(ICraftingService.class);
+        ICraftingSubmitResult submitResult = craftingService.submitJob(job, requester, targetCpu, false, this.source);
+        if (!submitResult.successful()) {
+            calculationNotSuccessful = true;
+            fireEvent(true, submitResult.errorCode().toString());
+            return;
+        }
+
+        this.jobLink = submitResult.link();
+        this.futureJob = null;
+        setStartedCrafting();
+        prepareCPUAndStatus(craftingService);
+    }
+
+    public void jobStateChanged() {
+        ICraftingLink jobLink = this.jobLink;
+        if (jobLink == null) {
+            fireEvent(true, StatusConstants.UNKNOWN_ERROR);
+            return;
+        }
+
+        if (jobLink.isCanceled() && !isJobCanceled) {
+            fireEvent(false, StatusConstants.JOB_CANCELED);
+            setJobCanceled();
+            return;
+        }
+
+        if (jobLink.isDone() && !isJobDone) {
+            fireEvent(false, StatusConstants.JOB_DONE);
+            setJobDone();
+        }
+    }
+
+    private void prepareCPUAndStatus(ICraftingService service) {
+        if (jobLink == null || jobStatus != null || !startedCrafting) {
+            return;
+        }
+        for (ICraftingCPU cpu : service.getCpus()) {
+            if (cpu instanceof CraftingCPUCluster cpuCluster) {
+                if (cpuCluster.craftingLogic.getLastLink() != null && cpuCluster.craftingLogic.getLastLink().getCraftingID().equals(jobLink.getCraftingID())) {
```

Comment:
Can we define a temporary variable here instead of copy paste same expr twice 😅

Reply:
Uh does that even work since the expression can change and the second expression is called later in the supplier?
Or do I overlook something here